### DISCUSSION TITLE
Confused with default ModuleName.Client

### DISCUSSION THREAD

**Discussion by jaeyson** [aj-foster/open-api-generator#72] [Q&A] [answered]
Hi!

I'm confused on how to use an HTTP client (e.g. `Req`) instead of the default `ExampleModule.Client`. After I generated it from the spec file (`ExampleModule.Client` is not defined) and wanted to use e.g. health check, what's the next step for using non default http client?

**Edit** Let me rephrase what I meant: after I use `mix api.gen ...` then I can edit the http client? What's the best course of action?

It seems that after I generate e.g. `mix api.gen...`, there's an implementation of the http client:

e.g.

```elixir
config :oapi_generator,
  default: [
    output: [
      base_module: ExampleModule,
      location: "lib/example_module",
      operation_subdirectory: "operations/",
      schema_subdirectory: "schemas/"      
    ]
  ]
end
```

```elixir
@default_client ExampleModule.Client

@doc """
Creates an analytics rule

When an analytics rule is created, we give it a name and describe the type, the source collections and the destination collection.
"""
@spec create_analytics_rule(ExampleModule.AnalyticsRuleSchema.t(), keyword) ::
        {:ok, ExampleModule.AnalyticsRuleSchema.t()}
        | {:error, ExampleModule.ApiResponse.t()}
def create_analytics_rule(body, opts \\ []) do
  client = opts[:client] || @default_client

  client.request(%{
    args: [body: body],
    call: {ExampleModule.Analytics, :create_analytics_rule},
    url: "/analytics/rules",
    body: body,
    method: :post,
    request: [{"application/json", {ExampleModule.AnalyticsRuleSchema, :t}}],
    response: [
      {201, {ExampleModule.AnalyticsRuleSchema, :t}},
      {400, {ExampleModule.ApiResponse, :t}}
    ],
    opts: opts
  })
end
```

Which from my understanding `client.request` is hardcoded, But I don't know how to use other adapter to bend according to its needs.

**Comment by aj-foster** [comment, marked as answer]
Hello there!

It looks like there's a naming conflict between the operations in your API's description and the default names used by the generator. But don't worry, because you can fix it.

The code generator doesn't want to force you to use a specific library for making HTTP requests, so it makes all of its function calls to a generic module like `MyApp.Client`. This is a module that you would implement, and it can pass on the request to any HTTP library you would like. [Here's an example](https://github.com/aj-foster/open-api-github/blob/main/lib/github/client.ex).

Unfortunately, it looks like the API description has some operations that would naturally be placed in a "Client" module. This is unusual to find in an API description, but obviously not impossible. Luckily there's a way for you to change the name of the module saved in the `@default_client` module attribute: `output.default_client`. For example:

```elixir
      config :oapi_generator, default: [
        output: [
          default_client: MyLib.MyClient
        ]
      ]
```

In this example, you would create the `MyClient` module and have it use `Req` for its requests.

**Reply by jaeyson** [reply]
Thank you for helping me out @aj-foster!
